### ISSUE TITLE
New globber walk

### ISSUE THREAD

**Issue by devlead** [cake-build/cake#377]
Upgrading a project to lates Cake I notices code like this won't work anymore
```csharp
Func<IFileSystemInfo, bool> exclude_node_modules =
fileSystemInfo=>!fileSystemInfo.Path.FullPath.StartsWith(
                "node_modules",
                StringComparison.OrdinalIgnoreCase);

var solutions = GetFiles("./**/*.sln", exclude_node_modules);
```
This was a hack to exclude certain folders from being traversed, i.e. because permissions, long paths etc.

Semantics in Globber changed ( https://github.com/cake-build/cake/commit/0e0c4784cbeea16caaa175ad7cf2dee4ec4b7721#diff-3d4e3c4984d72ab35ad74188651c2494R63 )  so it now will scan recursively first then use the filter predicate, before the predicate was sent to globber and used while traversing each level of directories/files.

Don't have a good solution to this yet, new way is much faster in most scenarios, but in the case finding only shallow files/dirs. in a deep folder structure and you can exclude known uninteresting bottomless pits (like node_modules i.e.), well then the old way is faster and won't break the build.

**Comment by RichiCoder1** [comment]
Completely unrelated: NPM 3+ allows for flat directories. It's beta, but its super stable near as I can tell and not having to deal with the infinite folders of doom helps a lot. There's even an awesome tool for upgrading it: https://github.com/felixrieseberg/npm-windows-upgrade

**Comment by patriksvensson** [comment]
Shouldn't be that difficult to add pre recursion filtering to the glob expression walker. I can take a quick look at this later this weekend.

**Comment by patriksvensson** [comment]
I'm assigning this to me. Going to give this a shot :smile:

**Comment by patriksvensson** [comment]
Ok, this is not as easy as it seems. There `predicate` part is not well defined. I always thought that the predicate was used to filter the results, but you actually want to use the predicate to decide what paths to traverse. These are two different things and I think we should decide upon what is the expected behavior of this. If it's used to choose paths to traverse, maybe `predicate` isn't the best name? Should this be documented better?

We should probably exclude personal bias for what we want to use this for ourselves and look at the bigger picture :smile:

**Comment by devlead** [comment]
What I'm trying to find a solution for is that if you have an "unsupported" directory structure you currently can't use any of the globber methods (`GetFiles`/`GetDirectories`/`CleanDirectories`/etc.) in your cake scrips, which is kinda sad as they're convenient aliases to use.

In some cases it might be using globber in a way it wasn't initially intended to be used, but for the "end user" it might not be obvious that i.e. `CleanDirectories(string.Concat(solutionDir, "/**/bin/", configuration));` will work first build (so it always works on CI which unless you cache does clean builds) but won't work build 2 because of "non" bin related artifacts generated by the build process.

For me if it's filtered before och after result created is an internal matter(alias user won't know the difference), they're cases to be made for both scenarios, and maybe we should have a totally different strategy or an  `overload` that let's script writer choose strategy. For small directory structures filtering after will be quicker and makes total sense, but if you have large directory structures (i.e.node_modules folders can easily be 10k folders and even more files.) and if you could hint the globber not to go down certain rabbit holes could really increase perf and avoid "long path issues".

Some things can be resolved by the ideas discussed in #261, but even if you can iterate 100k files&folders each build, it might be wasted CPU cycles if you already know certain paths are "dead ends".

**Comment by patriksvensson** [comment]
@devlead Closing this since the semantic change have been restored to how it was. The problem with long paths still exist, but should have it's own issue. There is currently PR #261 that tackles this but will probably need more work.
